confluent_kafka/src/confluent_kafka.c

```diff
@@ -47,6 +47,14 @@ PyObject *KafkaException;
  *        a new object if a rich error string is provided.
  *
  ****************************************************************************/
+
+#ifndef PyException_HEAD
+#define PyException_HEAD PyObject_HEAD PyObject *dict;\
```

Comment:
This looks fragile, subject to PyPy struct and type changes, and the lack of a public type/define in PyPy indicates (to me) that it shouldn't be used .

Do you have a link to where this struct is defined in PyPy?

Reply:
https://github.com/moderngl/moderngl/issues/311 is the only other place on github where `PyException_HEAD` is mentioned, and https://github.com/moderngl/moderngl/commit/540567f5f4c821001e70f06fa58695a48483dc10 is where it was removed from that project

https://foss.heptapod.net/search?search=PyException_HEAD&group_id=166&project_id=76 isnt mentioned at all in PyPy project.  I also grepped the code locally.

More oddly, `__suppress_context__` is also not present in PyPy source code.
Maybe it doesnt implement https://www.python.org/dev/peps/pep-0415/ ?

Anyway, on pypy3 7.3.7 on macos, I get an error running the tests on this branch.

```
tests/test_Admin.py::test_create_partitions_api python(19139,0x117e76600) malloc: Heap corruption detected, free list is damaged at 0x600001cc9ac0
*** Incorrect guard value: 175958608091839
python(19139,0x117e76600) malloc: *** set a breakpoint in malloc_error_break to debug
Fatal Python error: Aborted

Stack (most recent call first, approximate line numbers):
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/main.py", line 515 in gethookproxy
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/nodes.py", line 192 in ihook
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/runner.py", line 240 in call_runtest_hook
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/runner.py", line 212 in call_and_report
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/runner.py", line 114 in runtestprotocol
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/runner.py", line 106 in pytest_runtest_protocol
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_callers.py", line 9 in _multicall
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_manager.py", line 77 in _hookexec
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_hooks.py", line 244 in __call__
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/main.py", line 336 in pytest_runtestloop
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_callers.py", line 9 in _multicall
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_manager.py", line 77 in _hookexec
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_hooks.py", line 244 in __call__
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/main.py", line 319 in _main
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/main.py", line 256 in wrap_session
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_callers.py", line 9 in _multicall
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_manager.py", line 77 in _hookexec
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/pluggy/_hooks.py", line 244 in __call__
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/config/__init__.py", line 130 in main
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/site-packages/_pytest/config/__init__.py", line 178 in console_main
  File "/Users/jayvdb/inv/confluent-kafka-python/.tox/pypy3/bin/pytest", line 3 in <module>
  File "<builtin>/app_main.py", line 867 in execfile
  File "<builtin>/app_main.py", line 109 in run_toplevel
  File "<builtin>/app_main.py", line 607 in run_command_line
  File "<builtin>/app_main.py", line 944 in entry_point
  ```
I dont see similar errors on CPython 3.7.12

Reply:
If `test_Admin` and `tests/integration/` is skipped, I see four test failures:
```
FAILED tests/test_Consumer.py::test_basic_api - ValueError: Failed to create consumer: group.id must be set
FAILED tests/test_KafkaError.py::test_kafkaError_unknonw_error - TypeError: exceptions must derive from BaseException, not type
FAILED tests/test_Producer.py::test_basic_api - Failed: DID NOT RAISE <class 'TypeError'>
FAILED tests/test_docs.py::test_verify_docs - assert 211 == 0
```